### DISCUSSION TITLE
Error in untested AwkwardForth case

### DISCUSSION THREAD

**Discussion by renyhp** [scikit-hep/uproot5#826] [Q&A] [answered]
Hi, I have just installed uproot, but I can't get it to read a tree.

My code is very simple:

```python3
import uproot
f = uproot.reading.open("myfile.root")
t = f.get("aTree")
t['TheOnlyBranch'].array()
```

This just crashes with this stacktrace:
```text
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/tmp/venv/lib/python3.10/site-packages/uproot/behaviors/TBranch.py", line 834, in arrays
    _ranges_or_baskets_to_arrays(
  File "/tmp/venv/lib/python3.10/site-packages/uproot/behaviors/TBranch.py", line 3124, in _ranges_or_baskets_to_arrays
    uproot.source.futures.delayed_raise(*obj)
  File "/tmp/venv/lib/python3.10/site-packages/uproot/source/futures.py", line 36, in delayed_raise
    raise exception_value.with_traceback(traceback)
  File "/tmp/venv/lib/python3.10/site-packages/uproot/behaviors/TBranch.py", line 3066, in basket_to_array
    basket_arrays[basket.basket_num] = interpretation.basket_array(
  File "/tmp/venv/lib/python3.10/site-packages/uproot/interpretation/objects.py", line 139, in basket_array
    output = self.basket_array_forth(
  File "/tmp/venv/lib/python3.10/site-packages/uproot/interpretation/objects.py", line 206, in basket_array_forth
    output = self._discover_forth(
  File "/tmp/venv/lib/python3.10/site-packages/uproot/interpretation/objects.py", line 286, in _discover_forth
    output[i] = self._model.read(
  File "/tmp/venv/lib/python3.10/site-packages/uproot/containers.py", line 609, in read
    return uproot.deserialization.read_object_any(
  File "/tmp/venv/lib/python3.10/site-packages/uproot/deserialization.py", line 274, in read_object_any
    obj = cls.read(chunk, cursor, context, file, selffile, parent)
  File "/tmp/venv/lib/python3.10/site-packages/uproot/model.py", line 1372, in read
    versioned_cls.read(
  File "/tmp/venv/lib/python3.10/site-packages/uproot/model.py", line 867, in read
    self.read_members(chunk, cursor, context, file)
  File "<dynamic>", line 567, in read_members
KeyError: dtype('int8')
```

The branch does contain subbranches, some of them are of type `int8` and `uint8`.

Am I doing anything wrong or is this a bug to be reported?

**Comment by Moelf** [comment]
do you have a sample file?

**Reply by renyhp** [reply]
Sorry, I'm afraid I cannot share it. Any investigation I can do myself?

**Reply by tamasgal** [reply]
It's a bit difficult to help but you could e.g. show `t['TheOnlyBranch'].interpretation`

**Reply by jpivarski** [reply]
Does the branch have subbranches?

```python
t["TheOnlyBranch"].keys()
```

Maybe some subbranches are readable and others not. Trying individual subbranches can help you/us narrow down on what, exactly, is failing. (Branches of type `int8` and `uint8` should be perfectly fine: we have some of those in our test suite.)

If you're allowed to share the names and types of branches, with the data itself, then you can show us the result of

```python
t.show()
```

It's failing in `read_object_any`, which doesn't get invoked for simple branch types like `int8` and `uint8`. It's also trying to build an AwkwardForth VM, which only happens on a branch of type `AsObjects` (i.e. much more complex than numerical or even jagged/ragged arrays of numerical data).

After getting the KeyError, the class that was created to read this particular type of complex data structure should exist in `uproot.dynamic`. If you can find the relevant one for your data, print that class object's `class_code`:

```python
print(uproot.dynamic.WhateverYourClassIsNamed.class_code)
```

If you can't find it in `uproot.dynamic`, it should also be in the `AsObjects` interpretation:

```python
print(branch.interpretation.model.class_code)
```

(Although if the class itself is inside of an STL vector or something, you'd need to dig through the `AsVector.values` to get to the class object itself, which is what has the `class_code` that you can print.)

The actual error happens on line 567 of this dynamically generated class's code. If you're allowed to share type information but not values, you can post that class code here; it's a strict function of type, not any data values.

**Reply by renyhp** [reply]
Found it, thanks!

It is, in fact in one of the few subbranches that are complex objects. However this should only be holding numeric values, and a `TVector3`... 

Here is the class code (sorry for a bit of obfuscation...): [class_code.txt](https://github.com/scikit-hep/uproot5/files/10758713/class_code.txt)

**Reply by jpivarski** [reply]
Using the class_code and line number from the stack trace, I found what is very likely the error.

#827 has a bug-fix. Would you be able to test it by installing Uproot from that git branch?

```bash
pip uninstall uproot   # or remove it from conda; whichever package manager you're using

git clone https://github.com/scikit-hep/uproot5.git
cd uproot5
git checkout jpivarski/fix-missing-dtype-in-forth-symbol_dict
pip install -e .

# try reading the "SubBranch" class that has TVector3, a few arrays, and lots of scalars...
```

If that works, you can pip uninstall this Uproot and reinstall it with pip or conda when the next patch release of Uproot goes out.

Thanks!

**Reply by renyhp** [reply]
Well, now that I'm trying to access the same branch, I get this error:

```text
Traceback (most recent call last):
  File "/tmp/uproot5/src/uproot/interpretation/objects.py", line 232, in basket_array_forth
    context["forth"].vm = awkward.forth.ForthMachine64(
ValueError: in AwkwardForth source code, line 17 col 1, output dtype not recognized:

    output node13-data 

(https://github.com/scikit-hep/awkward/blob/awkward-cpp-9/awkward-cpp/src/libawkward/forth/ForthMachine.cpp#L1985)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/tmp/uproot5/src/uproot/behaviors/TBranch.py", line 1818, in array
    _ranges_or_baskets_to_arrays(
  File "/tmp/uproot5/src/uproot/behaviors/TBranch.py", line 3124, in _ranges_or_baskets_to_arrays
    uproot.source.futures.delayed_raise(*obj)
  File "/tmp/uproot5/src/uproot/source/futures.py", line 36, in delayed_raise
    raise exception_value.with_traceback(traceback)
  File "/tmp/uproot5/src/uproot/behaviors/TBranch.py", line 3066, in basket_to_array
    basket_arrays[basket.basket_num] = interpretation.basket_array(
  File "/tmp/uproot5/src/uproot/interpretation/objects.py", line 139, in basket_array
    output = self.basket_array_forth(
  File "/tmp/uproot5/src/uproot/interpretation/objects.py", line 236, in basket_array_forth
    raise type(err)(
ValueError: in AwkwardForth source code, line 17 col 1, output dtype not recognized:

    output node13-data 

(https://github.com/scikit-hep/awkward/blob/awkward-cpp-9/awkward-cpp/src/libawkward/forth/ForthMachine.cpp#L1985)

Forth code generated for this data type:

input stream
    input byteoffsets
    input bytestops
    output node0-data float64
output node1-data float64
output node2-data float64
output node3-data float64
output node4-data float64
output node5-data float64
output node6-data float64
output node7-data float64
output node8-data float32
output node9-data float32
output node10-data float32
output node11-data float32
output node12-data float32
output node13-data None
output node14-data bool
output node15-data bool
output node16-data bool
output node17-data bool
output node18-data float32
output node19-data float32
output node20-data float64
output node21-data float64
output node22-data float64
output node23-data bool
output node24-data bool
output node25-data float64
output node26-data float64
output node27-data None
output node28-data float32
output node29-data float32
output node30-data float32
output node31-data None
output node32-data float32
output node33-data bool
output node34-data bool
output node35-data None
output node36-data None
output node37-data float32
output node38-offsets int64
output node39-data float64
output node40-data float64
output node41-data float64
output node42-data float64
output node43-data float64
output node44-data float32
output node45-data float32
output node46-data float32
output node47-data float32
output node48-data float32
output node49-data None
output node50-data bool
output node51-data bool
output node52-data bool
output node53-data bool
output node54-data float32
output node55-data float32
output node56-data float64
output node57-data float64
output node58-data float64
output node59-data bool
output node60-data bool
output node61-data float64
output node62-data float64
output node63-data None
output node64-data float32
output node65-data float32
output node66-data float32
output node67-data None
output node68-data float32
output node69-data bool
output node70-data bool
output node71-data None
output node72-data None
output node73-data int32
output node74-offsets int64
output node75-data None
output node76-offsets int64
output node77-data float32
output node78-offsets int64
output node79-data float32
output node80-offsets int64
output node81-data float32
output node82-offsets int64
output node83-data float32
output node84-offsets int64
output node85-data float32
output node86-offsets int64

    0 node38-offsets <- stack
0 node74-offsets <- stack
0 node76-offsets <- stack
0 node78-offsets <- stack
0 node80-offsets <- stack
0 node82-offsets <- stack
0 node84-offsets <- stack
0 node86-offsets <- stack

    0 do
    byteoffsets I-> stack
    stream seek
    0 stream skip 
6 stream skip
10 stream skip 
0 stream skip 
6 stream skip
10 stream skip 
stream !d-> node0-data
stream !d-> node1-data
stream !d-> node2-data
stream !d-> node3-data
stream !d-> node4-data
stream !d-> node5-data
stream !d-> node6-data
stream !d-> node7-data
stream !f-> node8-data
stream !f-> node9-data
stream !f-> node10-data
stream !f-> node11-data
stream !f-> node12-data
stream !b-> node13-data
stream !?-> node14-data
stream !?-> node15-data
stream !?-> node16-data
stream !?-> node17-data
stream !f-> node18-data
stream !f-> node19-data
stream !d-> node20-data
stream !d-> node21-data
stream !d-> node22-data
stream !?-> node23-data
stream !?-> node24-data
stream !d-> node25-data
stream !d-> node26-data
stream !b-> node27-data
stream !f-> node28-data
stream !f-> node29-data
stream !f-> node30-data
stream !b-> node31-data
stream !f-> node32-data
stream !?-> node33-data
stream !?-> node34-data
stream !b-> node35-data
stream !b-> node36-data
5 dup node38-offsets +<- stack 
 stream #!f-> node37-data
stream !d-> node39-data
stream !d-> node40-data
stream !d-> node41-data
stream !d-> node42-data
stream !d-> node43-data
stream !f-> node44-data
stream !f-> node45-data
stream !f-> node46-data
stream !f-> node47-data
stream !f-> node48-data
stream !b-> node49-data
stream !?-> node50-data
stream !?-> node51-data
stream !?-> node52-data
stream !?-> node53-data
stream !f-> node54-data
stream !f-> node55-data
stream !d-> node56-data
stream !d-> node57-data
stream !d-> node58-data
stream !?-> node59-data
stream !?-> node60-data
stream !d-> node61-data
stream !d-> node62-data
stream !b-> node63-data
stream !f-> node64-data
stream !f-> node65-data
stream !f-> node66-data
stream !b-> node67-data
stream !f-> node68-data
stream !?-> node69-data
stream !?-> node70-data
stream !b-> node71-data
stream !b-> node72-data
2 dup node74-offsets +<- stack 
 stream #!i-> node73-data
2 dup node76-offsets +<- stack 
 stream #!b-> node75-data
2 dup node78-offsets +<- stack 
 stream #!f-> node77-data
2 dup node80-offsets +<- stack 
 stream #!f-> node79-data
2 dup node82-offsets +<- stack 
 stream #!f-> node81-data
2 dup node84-offsets +<- stack 
 stream #!f-> node83-data
2 dup node86-offsets +<- stack 
 stream #!f-> node85-data

    loop
    
```

**Reply by renyhp** [reply]
I tried patching `_awkward_forth.convert_dtype()` by adding

```python3
    elif format == "b":
        return "int8"
```

but now I get:

```text
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/tmp/uproot5/src/uproot/behaviors/TBranch.py", line 1818, in array
    _ranges_or_baskets_to_arrays(
  File "/tmp/uproot5/src/uproot/behaviors/TBranch.py", line 3124, in _ranges_or_baskets_to_arrays
    uproot.source.futures.delayed_raise(*obj)
  File "/tmp/uproot5/src/uproot/source/futures.py", line 36, in delayed_raise
    raise exception_value.with_traceback(traceback)
  File "/tmp/uproot5/src/uproot/behaviors/TBranch.py", line 3066, in basket_to_array
    basket_arrays[basket.basket_num] = interpretation.basket_array(
  File "/tmp/uproot5/src/uproot/interpretation/objects.py", line 139, in basket_array
    output = self.basket_array_forth(
  File "/tmp/uproot5/src/uproot/interpretation/objects.py", line 254, in basket_array_forth
    context["forth"].vm.resume()
ValueError: 'read beyond' in AwkwardForth runtime: tried to read beyond the end of an input
```

Using `library="np"` or `library="pd"` works, though!

**Reply by jpivarski** [reply]
Thanks for finding this other missing piece; I've added it to the PR.

`library="np"` and `library="pd"` both bypass the AwkwardForth reading, switching to a much slower (like, 400× slower) Python reader. It's possible for you to do that with `library="ak"` by setting

```python
t["your/branch"].interpretation._forth = False
```

before reading the array. This is not a public interface (see the underscore), so be aware that it's not officially supported—we want to get the AwkwardForth right.

You've encountered a case that we couldn't test for because of the range of ROOT formats we have available. It would be great to add your sample to our testing suite, but I know that's not something we can do.

How about this, can you print out

```python
t["your/branch"].debug_array(0).tobytes()
```

and

```python
t["your/branch"].typename
```

so I can try to step through it? I have the `class_code` and the Forth code. I'm a little suspicious of

```forth
    0 do
    byteoffsets I-> stack
    stream seek
    0 stream skip 
6 stream skip
10 stream skip 
0 stream skip 
6 stream skip
10 stream skip 
```

because you have only one `TVector3`, which has a 16 byte header to skip. Maybe the fact that your `SubBranch` is a `TObject` means there's another 4 or 6 bytes to skip, but not 16.

Another possibility: could you send me your file by email? I would keep it private.

**Reply by renyhp** [reply]
It might be relevant at this point to mention that this object is a custom one, so `t['TheOnlyBranch/SubBranch'].typename` is just something like `CustomObject*`.

I think I can provide you with a sample file by email, but it will take a while (maybe next week).

**Reply by jpivarski** [reply]
I know; I expected the `typename` to be `SubBranch` (because that's in the name of the class provided by `class_code`). I just wanted to make sure that it's not inside of an STL container, because that would change my interpretation of the bytes.

I can wait until next week.

I've half-convinced myself that the first 16 bytes is the `SubBranch`'s size header + `TObject` header and the next 16 bytes is the `TVector3`'s size header + `TObject` header, so it wouldn't explain why this is running off the end of the data stream.

**Reply by renyhp** [reply]
Writing here too for documentation: I have sent @jpivarski a confidential test file via email.

An identical tree with 0 entries fails with a different, more meaningful, error:

```text
ValueError: cannot produce Awkward Arrays for interpretation AsObjects(AsPointer(Model_SubBranch)) because

    arbitrary pointer

instead, try library="np" instead of library="ak" or globally set uproot.default_library
```

**Comment by jpivarski** [comment, marked as answer]
Thanks for (privately) sending the file! As you noticed when you sent it to me, empty TTrees fail with

```
ValueError: cannot produce Awkward Arrays for interpretation
AsObjects(AsPointer(Model_SubBranch)) because

     arbitrary pointer

instead, try library="np" instead of library="ak" or globally set
uproot.default_library
```

and that is the fundamental error. When executed in `library="np"` (arbitrary Python objects in NumPy arrays), it goes through [read_object_any](https://github.com/scikit-hep/uproot5/blob/c24159917ea5732ada3bb0b03615707823194197/src/uproot/deserialization.py#L184-L319), which deserializes a pointer type in C++, which could be a polymorphic class (which could be a union type in Awkward Array _if_ we knew the set of possible types, which we don't) and could have reference cycles (Awkward Arrays are strictly tree-like). The Python code in `library="np"` can handle those dynamic situations, but Awkward Array and AwkwardForth can't.

The AwkwardForth code that was generated is not correct because the `read_object_any` function doesn't generate any AwkwardForth code. But then again, there is no AwkwardForth code that would be correct, anyway.

So the issue here is just that the above error message didn't happen sooner. The zero-entry TTrees got to the point where this error is shown because they had nothing to deserialize and didn't error-out there. Sorry, but `library="np"` is your only option for this type of data. With PR #838, though, you'd get an error message saying that.

I'm deleting your file now, so that I don't mistakenly post it publicly in the future.
